includes/class-newspack-listings-featured.php

```diff
@@ -0,0 +1,536 @@
+				'type'              => 'boolean',
+			],
+			'priority' => [
+				'auth_callback'     => function() {
+					return current_user_can( 'edit_posts' );
+				},
+				'default'           => 5,
+				'description'       => __( 'What priority should this featured listing have (1–10)?', 'newspack-gdg' ),
+				'sanitize_callback' => 'absint',
+				'single'            => true,
+				'show_in_rest'      => true,
+				'type'              => 'integer',
+			],
+			'expires'  => [
+				'auth_callback'     => function() {
+					return current_user_can( 'edit_posts' );
+				},
+				'default'           => '',
+				'description'       => __( 'When should the listing’s featured status expire?', 'newspack-gdg' ),
+				'sanitize_callback' => 'sanitize_text_field',
+				'single'            => true,
+				'show_in_rest'      => true,
+				'type'              => 'string',
+			],
+		];
+
+		foreach ( Core::NEWSPACK_LISTINGS_POST_TYPES as $post_type => $post_type_slug ) {
+			foreach ( $meta_config as $key => $settings ) {
+				$settings['object_subtype'] = $post_type_slug;
+				register_meta(
+					'post',
+					self::META_KEYS[ $key ],
+					$settings
+				);
+			}
+		}
+	}
+
+	/**
+	 * Is the given/current post a featured post?
+	 *
+	 * @param int $post_id Post ID. If none given, use the current post ID.
+	 *
+	 * @return boolean True if the listing is currently featured.
+	 */
+	public static function is_featured( $post_id = null ) {
+		if ( null === $post_id ) {
+			$post_id = get_the_ID();
+		}
+
+		return get_post_meta( $post_id, self::META_KEYS['featured'], true );
+	}
+
+	/**
+	 * Get the feature priority for the given/current post.
+	 *
+	 * @param int $post_id Post ID. If none given, use the current post ID.
+	 *
+	 * @return int Feature priority level (default = 5).
+	 *             Will be returned whether or not the listing is featured.
+	 */
+	public static function get_featured_priority( $post_id = null ) {
+		if ( null === $post_id ) {
+			$post_id = get_the_ID();
+		}
+
+		return get_post_meta( $post_id, self::META_KEYS['priority'], true );
+	}
+
+	/**
+	 * Get the expiration date string for the given/current post.
+	 *
+	 * @param int $post_id Post ID. If none given, use the current post ID.
+	 *
+	 * @return string Featured status expiration date, or an empty string if none.
+	 */
+	public static function get_featured_expiration( $post_id = null ) {
+		if ( null === $post_id ) {
+			$post_id = get_the_ID();
+		}
+
+		return get_post_meta( $post_id, self::META_KEYS['expires'], true );
+	}
+
+	/**
+	 * On save, duplicate the feature priority meta value to a query meta key if the the item is featured.
+	 * If the item is not featured, delete the query meta value.
+	 * This lets us query based only on this meta value instead of checking both feature status and priority.
+	 *
+	 * @param int     $post_id Post ID.
+	 * @param WP_Post $post Post object.
+	 */
+	public static function set_feature_priority( $post_id, $post ) {
+		if ( Core::is_listing() ) {
```

Comment:
In my testing, when saving a post, the post is not recognised as listing and hence the priority is not updated.

Reply:
Update! It seems this was inconsistent. `get_post_type()` apparently doesn't always return a post when used on the `save_post` hook? 8343ccc seems to fix the issue by passing the post type to the `is_listing` helper directly from the hook's `post` object.